### ISSUE TITLE
JavaScript renderer source code is unreadable

### ISSUE THREAD

**Issue by ahtcx** [Microsoft/AdaptiveCards#2385]
# Platform

* JavaScript

# Issue

The code is mostly bundles in a 6000+ line file which makes it terribly hard to read and visualise the structure.

**Comment by dclaux** [comment]
Yes, it's far from ideal right now. Eventually it will get better, but I have no bandwidth for now.

**Comment by matthidinger** [comment]
Hi @ahtcx, indeed if you're trying to just read the minified browser bundle it's gonna be pretty hard to follow. The bundle (adaptivecards.min.js) is generated from our TypeScript source code and then passed to WebPack which transpiles our ES6 source code/modules into "browser-ready" functions.

Thankfully, all major browsers have tooling to make browsing the "real" source code easier.

In Chrome, open the developer tools and go to the **Sources** tab, if the library is loaded then you should see an **AdaptiveCards** entry in the tree, and can open a folder **./src** which has all the TypeScript source files. You can even set breakpoints in there and thanks to the magic of source-maps everything should be good :)

![image](https://user-images.githubusercontent.com/1432195/53110629-38211b80-34f0-11e9-9288-6f51bd802376.png)

The same thing exists in Edge, just go to the **Debugger** tab and look under **webpack:///AdaptiveCards/...**

![image](https://user-images.githubusercontent.com/1432195/53110820-b41b6380-34f0-11e9-9c17-5bb13e47386b.png)

**Comment by matthidinger** [comment]
.....and after writing that I realized that even `card-elements.ts` is still a 6000 line file, so you knew all that :)

Indeed like @dclaux mentioned we're planning to separate things out a bit in the future. It's definitely on our radar so I'll close this issue for now, but please keep the feedback coming!

**Comment by dclaux** [comment]
The main problem we have right now is that we compile to individual JS files and webpack takes care of bundling them together. I have already spent quite a bit of time splitting the library into more files, but what happens in the end is webpack is unable to properly deal with circular references and the code errors out at runtime.

There is a solution though; we should build to an AMD module instead of CommonJS and use TypeScript namespaces instead of imports/exports. This results in TypeScript compiling to a single JS file that itself has issue dealing with circular references. We can then package that resulting file using webpack.

This will require some work and will likely break the public interface of the renderer (although not dramatically), but I think it's worth it because I do agree the code is too hard to read at the moment.
